java/src/main/java/org/whispersystems/libsignal/util/Base32Encoder.java

```diff
@@ -0,0 +1,445 @@
+package org.whispersystems.libsignal.util;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * A streaming Base32 encoder.
+ *
+ * Copied from org.bouncycastle.util.encoders.Base32Encoder, available at
+ * https://github.com/bcgit/bc-java/blob/master/core/src/main/java/org/bouncycastle/util/encoders/Base32Encoder.java.
+ */
```

Comment:
may need to include the literal bouncycastle license/copyright somewhere as well if we're directly including? https://github.com/bcgit/bc-java/blob/master/LICENSE.html  -- it looks MIT.

Reply:
possibly note why we are including/prefer this version and our expectations about it being constant time and the issue you're resolving?

Reply:
🤦  Your comment got me to take another look at this and I'm realizing now that Bouncy Castle's implementation isn't constant time, since it uses table lookups! Back to the drawing board I guess ...